pkg/live/planner/cluster.go

```diff
@@ -0,0 +1,289 @@
+	"github.com/GoogleContainerTools/kpt/pkg/live"
+	"github.com/GoogleContainerTools/kpt/pkg/status"
+	"github.com/GoogleContainerTools/kpt/thirdparty/cli-utils/apply"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/kubectl/pkg/cmd/util"
+	"sigs.k8s.io/cli-utils/pkg/apply/event"
+	"sigs.k8s.io/cli-utils/pkg/common"
+	"sigs.k8s.io/cli-utils/pkg/inventory"
+	"sigs.k8s.io/cli-utils/pkg/object"
+)
+
+func NewClusterPlanner(f util.Factory) (*ClusterPlanner, error) {
+	fetcher, err := NewResourceFetcher(f)
+	if err != nil {
+		return nil, err
+	}
+
+	invClient, err := inventory.NewClient(f, live.WrapInventoryObj, live.InvToUnstructuredFunc, inventory.StatusPolicyNone)
+	if err != nil {
+		return nil, err
+	}
+
+	statusPoller, err := status.NewStatusPoller(f)
+	if err != nil {
+		return nil, err
+	}
+
+	applier, err := apply.NewApplierBuilder().
+		WithFactory(f).
+		WithInventoryClient(invClient).
+		WithStatusPoller(statusPoller).
+		Build()
+	if err != nil {
+		return nil, err
+	}
+
+	return &ClusterPlanner{
+		applier:         applier,
+		resourceFetcher: fetcher,
+	}, nil
+}
+
+type Applier interface {
+	Run(ctx context.Context, invInfo inventory.Info, objects object.UnstructuredSet, options apply.ApplierOptions) <-chan event.Event
+}
+
+type ResourceFetcher interface {
+	FetchResource(ctx context.Context, id object.ObjMetadata) (*unstructured.Unstructured, bool, error)
+}
+
+type ClusterPlanner struct {
+	applier         Applier
+	resourceFetcher ResourceFetcher
+}
+
+type Options struct {
+	ServerSideOptions common.ServerSideOptions
+}
+
+func (r *ClusterPlanner) BuildPlan(ctx context.Context, inv inventory.Info, objects []*unstructured.Unstructured, o Options) (*planv1alpha1.Plan, error) {
+	actions, err := r.dryRunForPlan(ctx, inv, objects, o)
+	if err != nil {
+		return nil, err
+	}
+	return &planv1alpha1.Plan{
+		ResourceMeta: planv1alpha1.ResourceMeta,
+		Spec: planv1alpha1.PlanSpec{
+			Actions: actions,
+		},
+	}, nil
+}
+
+func (r *ClusterPlanner) dryRunForPlan(ctx context.Context, inv inventory.Info,
+	objects []*unstructured.Unstructured, o Options) ([]planv1alpha1.Action, error) {
+
+	eventCh := r.applier.Run(ctx, inv, objects, apply.ApplierOptions{
+		DryRunStrategy:    common.DryRunServer,
+		ServerSideOptions: o.ServerSideOptions,
+	})
+
+	var actions []planv1alpha1.Action
+	var err error
+	for e := range eventCh {
+		if e.Type == event.InitType {
+			// This event includes all resources that will be applied, pruned or deleted, so
+			// we make sure we fetch all the resources from the cluster.
+			// TODO: See if we can update the actuation library to provide the pre-actuation
+			// versions of the resources as part of the regular run. This solution is not great
+			// as fetching all resources will take time.
```

Comment:
Agreed - it also means that we risk not getting a consistent view if there are concurrent changes.